src/blockrelay/mempool_sync.cpp

```diff
@@ -0,0 +1,444 @@
+CMempoolSync::CMempoolSync(std::vector<uint256> mempoolTxHashes,
+    uint64_t nReceiverMemPoolTx,
+    uint64_t nSenderMempoolPlusBlock,
+    uint64_t shorttxidk0,
+    uint64_t shorttxidk1,
+    uint64_t _version)
+{
+    uint64_t grapheneSetVersion = CMempoolSync::GetGrapheneSetVersion(version);
+    version = _version;
+    nSenderMempoolTxs = 0;
+    nSenderMempoolTxs = mempoolTxHashes.size();
+
+    pGrapheneSet = std::make_shared<CGrapheneSet>(CGrapheneSet(nReceiverMemPoolTx, nSenderMempoolPlusBlock,
+        mempoolTxHashes, shorttxidk0, shorttxidk1, grapheneSetVersion, IBLT_ENTROPY, COMPUTE_OPTIMIZED, false));
+}
+
+CMempoolSync::~CMempoolSync() { pGrapheneSet = nullptr; }
+bool HandleMempoolSyncRequest(CDataStream &vRecv, CNode *pfrom)
+{
+    LOG(MPOOLSYNC, "Handling mempool sync request from peer %s\n", pfrom->GetLogName());
+    CMempoolSyncInfo mempoolinfo;
+    CInv inv;
+    vRecv >> inv >> mempoolinfo;
+
+    // Message consistency checking
+    if (!(inv.type == MSG_MEMPOOLSYNC))
+    {
+        dosMan.Misbehaving(pfrom, 100);
+        return error("invalid GET_MEMPOOLSYNC message type=%u\n", inv.type);
+    }
+
+    // Requester should only contact peers that support mempool sync
+    if (!syncMempoolWithPeers.Value())
+    {
+        dosMan.Misbehaving(pfrom, 100);
+        return error("Mempool sync requested from peer %s but not supported\n", pfrom->GetLogName());
+    }
+
+    // Requester must limit request frequency
+    if (mempoolSyncResponded.count(pfrom) > 0 &&
+        std::chrono::duration_cast<std::chrono::microseconds>(
+            std::chrono::high_resolution_clock::now() - mempoolSyncResponded[pfrom].lastUpdated)
+                .count() < MEMPOOLSYNC_FREQ_US - MEMPOOLSYNC_FREQ_GRACE_US)
+    {
+        dosMan.Misbehaving(pfrom, 100);
+        return error("Mempool sync requested less than %d mu seconds ago from peer %s\n", MEMPOOLSYNC_FREQ_US,
+            pfrom->GetLogName());
+    }
+
+    // Record request
+    mempoolSyncResponded[pfrom] = CMempoolSyncState(
+        std::chrono::high_resolution_clock::now(), mempoolinfo.shorttxidk0, mempoolinfo.shorttxidk1, false);
+
+    if (inv.type == MSG_MEMPOOLSYNC)
+    {
+        LOG(MPOOLSYNC, "Mempool currently holds %d transactions\n", mempool.size());
+
+        std::vector<uint256> mempoolTxHashes;
+        // cycle through mempool txs in order of ancestor_score
+        {
+            READLOCK(mempool.cs_txmempool);
+
+            int64_t nRemainingMempoolBytes = mempoolinfo.nRemainingMempoolBytes;
+            typename CTxMemPool::indexed_transaction_set::index<ancestor_score>::type::iterator it =
+                mempool.mapTx.get<ancestor_score>().begin();
+            for (; it != mempool.mapTx.get<ancestor_score>().end() && nRemainingMempoolBytes > 0; ++it)
+            {
+                size_t nTxSize = it->GetTx().GetTxSize();
+                int64_t nFee = it->GetFee();
+                uint64_t nSatoshiPerK = (uint64_t)(1000 * nFee / nTxSize);
+
```

Comment:
I think you should create a CFeeRate object here found in amount.h, to calculate nsatoshiperk,  then do a object.GetFeePerK().  Problem here is that what happens if nTxSize is zero, the CFeeRate class will already take care of that.

Reply:
Done.